src/module.c

```diff
@@ -464,6 +465,149 @@ int GetSingleDocumentCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int
+  if (retVal < 0) {
+    RedisModule_ReplyWithError(ctx, error);
+  } else {
+    RedisModule_ReplyWithLongLong(ctx, retVal);
+  }
+
+  return REDISMODULE_OK;
+}
+
+int DictAddCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+  char *err = NULL;
+  if (argc < 3) {
+    return RedisModule_WrongArity(ctx);
+  }
+
+  const char *dictName = RedisModule_StringPtrLen(argv[1], NULL);
+
+  char *error;
+  int retVal = SpellCheck_DictAdd(ctx, dictName, argv + 2, argc - 2, &error);
+  if (retVal < 0) {
+    RedisModule_ReplyWithError(ctx, error);
+  } else {
+    RedisModule_ReplyWithLongLong(ctx, retVal);
+  }
+
+  return REDISMODULE_OK;
+}
+
+int SpellCheckCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+#define DICT_INITIAL_ZISE 5
+#define DEFAULT_LEV_DISTANCE 1
+#define MAX_LEV_DISTANCE 100
+#define str(s) #s
+  char *err = NULL;
+  if (argc < 3) {
+    return RedisModule_WrongArity(ctx);
+  }
+
+  RedisModule_AutoMemory(ctx);
+  RedisSearchCtx *sctx = NewSearchCtx(ctx, argv[1]);
+  if (sctx == NULL) {
+    return RedisModule_ReplyWithError(ctx, "Unknown Index name");
+  }
+
+  size_t len;
+  const char *rawQuery = RedisModule_StringPtrLen(argv[2], &len);
+  QueryParseCtx *q = NewQueryParseCtx(sctx, rawQuery, len, NULL);
+  if (!q) {
+    SearchCtx_Free(sctx);
+    return RedisModule_ReplyWithError(ctx, "Error parsing query");
+  }
+
+  char **includeDict = array_new(char *, DICT_INITIAL_ZISE);
+  char **excludeDict = array_new(char *, DICT_INITIAL_ZISE);
+
+  if (!Query_Parse(q, &err)) {
+
+    if (err) {
+      RedisModule_Log(ctx, "debug", "Error parsing query: %s", err);
+      RedisModule_ReplyWithError(ctx, err);
+      ERR_FREE(err);
+    } else {
+      /* Simulate an empty response - this means an empty query */
+      RedisModule_ReplyWithArray(ctx, 1);
+      RedisModule_ReplyWithLongLong(ctx, 0);
+    }
+    goto end;
+  }
+
+  int distanceArgPos = 0;
+  long long distance = DEFAULT_LEV_DISTANCE;
+  if ((distanceArgPos = RMUtil_ArgExists("DISTANCE", argv, argc, 0))) {
+    if (distanceArgPos + 1 >= argc) {
+      RedisModule_ReplyWithError(ctx, "DISTANCE arg is given but no DISTANCE comes after");
+      goto end;
+    }
+    if (RedisModule_StringToLongLong(argv[distanceArgPos + 1], &distance) != REDISMODULE_OK ||
+        distance < 1 || distance > MAX_LEV_DISTANCE) {
+      RedisModule_ReplyWithError(
+          ctx, "bad distance given, distance must be a natural number between 1 to " str(
+                   MAX_LEV_DISTANCE));
+      goto end;
+    }
+  }
+
+  int nextPos = 0;
+  while ((nextPos = RMUtil_ArgExists("TERMS", argv, argc, nextPos + 1))) {
+    if (nextPos + 2 >= argc) {
+      RedisModule_ReplyWithError(ctx, "TERM arg is given but no TERM params comes after");
+      goto end;
+    }
+    char *operation = (char *)RedisModule_StringPtrLen(argv[nextPos + 1], NULL);
+    char *dictName = (char *)RedisModule_StringPtrLen(argv[nextPos + 2], NULL);
```

Comment:
const char. Why remove the constness? maybe you can remove it when adding to the array.